### PR TITLE
Rounding of Taxes and Transfers

### LINKED ISSUE #88: Rounding is missing for several taxes / transfers
### Bug description

Right now there are rounding functions in the code. We want to be as exact as possible and don't loose any information.

For example:
https://github.com/iza-institute-of-labor-economics/gettsim/blob/1b59c617b6b0c6f4acf845128988b60e207bc028/gettsim/taxes/calc_taxes.py#L138

### TODO:
- [ ] Remove all rounding functions.

### PR DESCRIPTION
**Pull request by MImmesberger**
Closes #88 

### Todo

- [ ] Wohngeld in steps of 5€.
- [x] Income deduction from Kinderzuschlag in steps of 10€ until 07/2019.
- [x] Vorsorgeaufwendungen are rounded up.
- [x] Income tax is rounded down.
- [x] Unterhaltsvorschuss is rounded up.

### CODE REVIEW

**Comment by MImmesberger** [comment]
The PR is ready for review.
1. I couldn't find the legal basis for Wohngeld being payed out in 5€ steps, so I did not implement it. See discussion in #88 
2. The rounding of zu versteuerndes Einkommen includes an additive part (e.g. round down to a number divisible by 54 and then add 18€). A summary is [here](https://www.bmf-steuerrechner.de/ekst/rundungsvorschrift.xhtml#:~:text=Für%20die%20Einkommensteuerformel%202002%2F2003,um%2018%20€%20zu%20erhöhen.). Because this wasn't possible with the current machinery, I added an "add" key to the rounding specification that adds a float after the rounding step.

**Comment by hmgaudecker** [comment]
> I couldn't find the legal basis for Wohngeld being payed out in 5€ steps, so I did not implement it. See discussion in

Yes, let's leave it at the nearest Euro!

File: src/_gettsim/interface.py [hunk 1]
```diff
@@ -604,22 +604,23 @@ def _add_rounding_to_functions(functions, params):
...
                 base=rounding_spec["base"],
                 direction=rounding_spec["direction"],
+                add=rounding_spec["add"] if "add" in rounding_spec else None,
             )(func)
 
     return functions_new
 
 
-def _add_rounding_to_one_function(base, direction):
+def _add_rounding_to_one_function(base, direction, add):
```
Comment (hmgaudecker) [inline src/_gettsim/interface.py:613]: Do we have tests for this or is codecov's complaint a fluke?

And please add typing, something like

```python
base: float,
direction: Literal['up', 'down', 'nearest'']
add: float
```

Not sure about what you mean by "optional"? Surely, the function expects to be called with `add` supplied. Do you mean that the typing should be `add: float|None` ?

However, given all the checks below, I'd think it'll be easier to just specify a 0 everywhere.

And we should be a bit more expressive in the name, maybe `to_add_after_rounding` ?

File: src/_gettsim/parameters/eink_st.yaml [hunk 2]
```diff
@@ -276,3 +276,21 @@ rounding:
       base: 1
       direction: down
       reference: § 32a Abs. 1 S. 6 EStG
+  zu_verst_eink_y_tu:
```
Comment (hmgaudecker) [inline src/_gettsim/parameters/eink_st.yaml:279]: Can you add an explanation of what the "add" does here? Seems a bit weird to me. Also, annual income was only recorded in steps of 54 Euros? Not 100 DM / 1.95583 = 51.13 EUR ?

File: src/_gettsim/parameters/unterhaltsvors.yaml [hunk 3]
```diff
@@ -49,3 +49,11 @@ altersgrenzen:
     1: 6
     2: 12
     3: 18
+rounding:
+  unterhaltsvors_m:
+    note:
+      en: Starting date unclear
```
Comment (hmgaudecker) [inline src/_gettsim/parameters/unterhaltsvors.yaml:55]: Why?

File: src/_gettsim_tests/test_rounding.py [hunk 4]
```diff
@@ -57,10 +63,15 @@ def test_func():
 
 
 @pytest.mark.parametrize(
-    "base, direction",
-    [(1, "upper"), ("0.1", "down"), (5, "closest")],
+    "base, direction, add",
+    [
+        (1, "upper", None),
```
Comment (hmgaudecker) [inline src/_gettsim_tests/test_rounding.py:68]:
```suggestion
        (1, "upper", 0),
```
See above, I don't think this should work.

File: src/_gettsim_tests/test_rounding.py [hunk 5]
```diff
@@ -131,6 +144,40 @@ def test_func(income):
     np.array_equal(calc_result["test_func"].values, np.array(input_values_exp_output))
 
 
+@pytest.mark.parametrize(
+    "base, direction, add, input_values, exp_output",
+    rounding_specs_and_exp_results_with_additive_part,
+)
+def test_rounding_with_additive_part(base, direction, add, input_values, exp_output):
```
Comment (hmgaudecker) [inline src/_gettsim_tests/test_rounding.py:151]: I don't think we need a separate test from `test_rounding` anymore, unless we leave the code as is.

**Review by hmgaudecker — COMMENTED** [review]
Very sweet! I think I would get rid of the case distinction for the `add` parameter (and rename it), just requiring that it is zero if not used.

File: src/_gettsim/interface.py (hunk 1 above)
Comment (MImmesberger) [reply src/_gettsim/interface.py]: > Do we have tests for this or is codecov's complaint a fluke?

If you are referring to the codecov report from above: this was a mistake on the codecov side. If not, I don't know what you are referring to.

> However, given all the checks below, I'd think it'll be easier to just specify a 0 everywhere.

I agree, I changed the default value from `None` to `0`:

```python
to_add_after_rounding=rounding_spec["to_add_after_rounding"]
                if "to_add_after_rounding" in rounding_spec
                else 0,
            )(func)
``

File: src/_gettsim/parameters/eink_st.yaml (hunk 2 above)
Comment (MImmesberger) [reply src/_gettsim/parameters/eink_st.yaml:279]: Ah that was a typo, should have been 36, not 54. I didn't convert currencies, but just took the rules from the EStG/ the Website of the ministry of finance:

> Für die Einkommensteuerformel 2002/2003 ist zur Ermittlung des abgerundeten zu versteuernden Einkommens (zvE) das zu versteuernde Einkommen auf den nächsten, durch 36 € ohne Rest teilbaren €-Betrag abzurunden, wenn es nicht bereits durch 36 ohne Rest teilbar ist und um 18 € zu erhöhen.

File: src/_gettsim/parameters/unterhaltsvors.yaml (hunk 3 above)
Comment (MImmesberger) [reply src/_gettsim/parameters/unterhaltsvors.yaml]: Looked into it and realized that the rounding rules were already there when Unterhaltsvorschuss was introduced in 1980. Updated the file accordingly.

File: src/_gettsim_tests/test_rounding.py (hunk 5 above)
Comment (MImmesberger) [reply src/_gettsim_tests/test_rounding.py]: I removed the `test_rounding_with_additive_part` and included the test in the existing machinery. I wanted to explicitly test the case where `to_add_after_rounding` is not specified in the dict, so I edited the part where this is relevant in the original test.

**Comment by MImmesberger** [comment]
Thank you for the comments! I hope I addressed all your concerns.

**Comment by MImmesberger** [comment]
Sorry, I still have trouble interpreting codecov outputs:

From the output I figured that `_add_rounding_to_one_function` is not properly covered by a test (only implicitly because it is used in the wrapper `_add_rounding_to_functions` which is tested). But testing directly via `test_rounding_callable` doesn't help.

Maybe @lars-reimann knows whats going on?

**Comment by hmgaudecker** [comment]
Oh, codecov happens to be off every now and then. If you are sure it is being run, that's fine!

File: src/_gettsim/parameters/eink_st.yaml (hunk 2 above)
Comment (hmgaudecker) [reply src/_gettsim/parameters/eink_st.yaml:279]: Makes sense for Euro-times. But the 54 before that?

**Review by hmgaudecker — APPROVED** [review]
Looks great, just one tiny comment left.

File: src/_gettsim/parameters/eink_st.yaml (hunk 2 above)
Comment (MImmesberger) [reply src/_gettsim/parameters/eink_st.yaml:279]: This is a rule in the EStG as well:

> Für Einkommensteuerformeln 1981 bis 2001 ist zur Ermittlung des abgerundeten zu versteuernden Einkommens (zvE) das zu versteuernde Einkommen auf den nächsten, durch 54 ohne Rest teilbaren DM-Betrag abzurunden, wenn es nicht bereits durch 54 ohne Rest teilbar ist. In 2001 ist dieser ermittelte Wert um 27 DM zu erhöhen.

Comment (hmgaudecker) [reply src/_gettsim/parameters/eink_st.yaml:279]: Oh, I am afraid you will have to convert that, then. We were thinking to have a possibility to specify unit=DM, but IIRC we never got around to that, yet?

Comment (MImmesberger) [reply src/_gettsim/parameters/eink_st.yaml:279]: Ah, I thought we calculate everything in DM before 2002. 

I converted the DM values to EUR.

Comment (hmgaudecker) [reply src/_gettsim/parameters/eink_st.yaml:279]: That would be the goal eventually, but so far, we did not get around to it!

**Pull request merged**